File: 480-sliding-window-median/480-sliding-window-median.cpp
class Solution {
public:
    vector<double> medianSlidingWindow(vector<int>& nums, int k) {
        auto gt_cmp = [&nums] (int i, int j) { 
            return nums[i] != nums[j] ? nums[i] > nums[j] : i < j; 
        };
        auto lt_cmp = [&nums] (int i, int j) { 
            return nums[i] != nums[j] ? nums[i] < nums[j] : i < j; 
        };
        
        set<int, decltype(gt_cmp)> left(gt_cmp);
        set<int, decltype(lt_cmp)> right(lt_cmp);
        
        function<double()> median;
        
        if (k & 1) {
            median = [&right, &nums] { return (double) nums[*right.begin()]; };
        } else {
            median = [&left, &right, &nums] { return ((double) nums[*left.begin()] + nums[*right.begin()]) / 2; };
        }
        
        auto balance = [&left, &right] () {
            while (left.size() > right.size()) {
                right.emplace(*left.begin());
                left.erase(left.begin());
            }
        };
        
        int n = nums.size();
        vector<double> ans(n - k + 1);
        
        for (int i = 0; i < k; ++i) left.emplace(i);
        balance(); ans[0] = median();
        
        for (int i = k; i < n; ++i) {
            if (!left.erase(i - k)) right.erase(i - k);
            right.emplace(i), left.emplace(*right.begin()), right.erase(right.begin());
            
            balance(); ans[i - k + 1] = median();
        }
        
        return ans;
    }
};

File: 480-sliding-window-median/480-sliding-window-median.java
class Solution {
    public double[] medianSlidingWindow(int[] nums, int k) {
        Comparator<Integer> comparator = new Comparator<>(){
             public int compare(Integer a, Integer b) {
                if (nums[a] != nums[b]) {
                    return Integer.compare(nums[a], nums[b]);
                }
                else {
                    return a - b;
                }
            } 
        };
        TreeSet<Integer> left = new TreeSet<>(comparator.reversed());
        TreeSet<Integer> right = new TreeSet<>(comparator);
        double[] res = new double[nums.length - k + 1];

        for (int i = 0; i < k; i++) {
            left.add(i);
        }
        balance(left, right);
        res[0] = getMedian(k, nums, left, right);

        int r = 1;
        for (int i = k; i < nums.length; i++) {
            if(!left.remove(i - k)) {
                right.remove(i - k);
            }
            right.add(i); 
            left.add(right.pollFirst());
            balance(left, right); 
            res[r] = getMedian(k, nums, left, right);
            r++;
        }

        return res;
    }
    
    private void balance(TreeSet<Integer> left, TreeSet<Integer> right) {
        while (left.size() > right.size()) {
            right.add(left.pollFirst());
        }
    }
    
    private double getMedian(int k, int[] nums, TreeSet<Integer> left, TreeSet<Integer> right) {
        if (k % 2 == 0) {
            return ((double) nums[left.first()] + nums[right.first()]) / 2;
        }
        else {
            return (double) nums[right.first()];
        }
    }
}